dev-guide/api-conventions.md

```diff
@@ -129,6 +129,139 @@ By providing quality documentation within the API itself, a number of generated
+declarative choice.
+
+#### What is a Discriminated Union?
+
+A discriminated union is a structure within the API that closely resembles a union type.
+A number of fields exist within the structure and we are expecting the user to configure precisely one of the fields.
+
+In particular, for a discriminated union, an extra field exists which allows the user to declaratively state which of particular fields they are configuring.
+
+We use discriminated unions in Kubernetes APIs so that we force the user to make a choice.
+We do not want our code to guess what the user meant, they should tell us which of the choices they made using the discriminant.
+
+#### Writing a union in Go
+
+Union types in Go have some additional helper tags which signify how the structure should be handled to consumers.
+Below is an example based on the idea of platform types.
+
+```go
+// MyPlatformConfig is a discriminated union of platform specific configuration.
+// It has a +union tag which informs consumers that this is expected to be a union type.
+// +union
+type MyPlatformConfig struct {
+  // PlatformType is the unions discriminator.
+  // Users are expected to set this value to the name of the platform.
+  // The value of this field should match, or be able to be mapped to exactly one field
+  // in the union structure.
+  // It has a +unionDiscriminator tag to inform consumers that this is the discriminator field.
+  // The field should be an enum type, so you may also need an enum tag.
+  // The field is required, though a future KEP will default this field if not set, at which point
+  // it can be made optional. See future section below.
+  // +unionDiscriminator
+  // +kubebuilder:validation:Enum:="AWS";"Azure";"GCP"
+  // +kubebuilder:validation:Required
+  PlatformType string `json:"platformType,omitempty"`
+
+  // AWS is the AWS configuration.
+  // All structures within the union must be optional and pointers.
+  // +optional.
+  AWS *MyAWSConfig `json:"aws,omitempty"`
+
+  // Azure is the Azure configuration.
+  // All structures within the union must be optional and pointers.
+  // +optional.
+  Azure *MyAzureConfig `json:"azure,omitempty"`
+
+  // GCP is the GCP configuration.
+  // All structures within the union must be optional and pointers.
+  // +optional.
+  GCP *MyGCPConfig `json:"gcp,omitempty"`
+}
+```
+
+The discriminator here allows the consumer to determine which of the configuration structures they should be consuming, AWS, Azure or GCP.
+
+Important to note:
+* All structs within the union **MUST** be pointers
+* All structs within the union **MUST** be optional
+* The discriminant should be required, but will be defaulted by the API server in a future Kubernetes version, at which point it can be optional
+* The discriminant **MUST** be a string (or string alias) type
+* If a discriminant value does not match the json tag of the struct to which it refers, a mapping can be made using an additional tag.
```

Comment:
Let's require it to match.

Reply:
On this one, we have a few examples where these match modulo a strings.ToLower, but don't necessarily match verbatim. For example we have some platform specific stuff, `platformType: AWS` and then the union field is `aws`. Do we think that is acceptable so that the discriminant can remain as PascalCase even though the field names are camelCase?

Reply:
In fact I've done this in the golang example I provided, should it actually be `// +kubebuilder:validation:Enum:="aws";"azure";"gcp` instead?

Reply:
Enumerations should be CamelCase in the API.  The upstream style guide will also be CamelCase.